## NumPy versus pandas

In [1]:
import pandas as pd
df = pd.read_csv("interviews.csv")

> Pandas is column-oriented

In [3]:
# Iterating pandas DataFrame by column
import time
start = time.time()
for col in df.columns:
    for item in df[col]:
        pass
print(time.time() - start, 'seconds')

0.012938976287841797 seconds


In [4]:
# Iterating pandas DataFrame by row
n_rows = len(df)
start = time.time()
for i in range(n_rows):
    for item in df.iloc[i]:
        pass
print(time.time() - start, 'seconds')

0.2445828914642334 seconds


> NumPy

In [5]:
df_np = df.to_numpy()
n_rows, n_cols = df_np.shape

In [6]:
# Iterating Numpy ndarray by column
start = time.time()
for j in range(n_cols):
    for item in df_np[:, j]:
        pass
print(time.time() - start, 'seconds')

0.004641056060791016 seconds


In [7]:
# Iterating Numpy ndarray by row
start = time.time()
for i in range(n_rows):
    for item in df_np[i]:
        pass
print(time.time() - start, 'seconds')

0.01610279083251953 seconds


Faster than 0.2445828914642334 seconds.

#### Reference:
- Huyen, Chip (2022), Designing Machine Learning Systems: An Iterative Process for Production-Ready Applications, O’Reilly.